File: dbm-ui/backend/ticket/builders/mysql/mysql_proxy_upgrade.py
# -*- coding: utf-8 -*-
"""
TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-DB管理系统(BlueKing-BK-DBM) available.
Copyright (C) 2017-2023 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at https://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""

from django.utils.translation import gettext_lazy as _
from rest_framework import serializers

from backend.flow.engine.controller.mysql import MySQLController
from backend.ticket import builders
from backend.ticket.builders.common.base import DisplayInfoSerializer
from backend.ticket.builders.mysql.base import BaseMySQLTicketFlowBuilder, MySQLBaseOperateDetailSerializer
from backend.ticket.constants import TicketType


class MysqlProxyUpgradeDetailSerializer(MySQLBaseOperateDetailSerializer):
    class InfoSerializer(DisplayInfoSerializer):
        cluster_ids = serializers.ListField(help_text=_("集群ID列表"), child=serializers.IntegerField(), min_length=1)
        pkg_id = serializers.IntegerField(help_text=_("目标版本包ID"))

    infos = serializers.ListField(help_text=_("单据信息"), child=InfoSerializer())


class MysqlProxyUpgradeParamBuilder(builders.FlowParamBuilder):
    controller = MySQLController.mysql_proxy_upgrade_scene


@builders.BuilderFactory.register(TicketType.MYSQL_PROXY_UPGRADE)
class MysqlProxyUpgradeFlowBuilder(BaseMySQLTicketFlowBuilder):
    serializer = MysqlProxyUpgradeDetailSerializer
    inner_flow_builder = MysqlProxyUpgradeParamBuilder

File: dbm-services/mysql/db-tools/dbactuator/internal/subcmd/proxycmd/mysql_proxy_upgrade.go
// Package proxycmd TODO
/*
 * TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-DB管理系统(BlueKing-BK-DBM) available.
 * Copyright (C) 2017-2023 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package proxycmd

import (
	"fmt"

	"github.com/spf13/cobra"

	"dbm-services/common/go-pubpkg/logger"
	"dbm-services/mysql/db-tools/dbactuator/internal/subcmd"
	"dbm-services/mysql/db-tools/dbactuator/pkg/components/mysql_proxy"
	"dbm-services/mysql/db-tools/dbactuator/pkg/util"
)

// MySQLProxyUpgradeAct TODO
type MySQLProxyUpgradeAct struct {
	*subcmd.BaseOptions
	Service mysql_proxy.UpgradeProxyComp
}

// NewMySQLProxyUpgradeAct TODO
func NewMySQLProxyUpgradeAct() *cobra.Command {
	act := MySQLProxyUpgradeAct{
		BaseOptions: subcmd.GBaseOptions,
	}
	cmd := &cobra.Command{
		Use:   "upgrade",
		Short: "升级mysql-proxy实例",
		Example: fmt.Sprintf("dbactuator proxy upgrade %s %s", subcmd.CmdBaseExampleStr,
			subcmd.ToPrettyJson(act.Service.Example())),
		Run: func(cmd *cobra.Command, args []string) {
			util.CheckErr(act.Init())
			util.CheckErr(act.Run())
		},
	}
	return cmd
}

// Init prepare run env
func (c *MySQLProxyUpgradeAct) Init() (err error) {
	if err = c.Deserialize(&c.Service.Params); err != nil {
		logger.Error("DeserializeAndValidate failed, %v", err)
		return err
	}
	c.Service.GeneralParam = subcmd.GeneralRuntimeParam
	return nil
}

// Run Command Run
func (c *MySQLProxyUpgradeAct) Run() error {
	steps := subcmd.Steps{
		{
			FunName: "初始化",
			Func:    c.Service.Init,
		},
		{
			FunName: "前置检查",
			Func:    c.Service.PreCheck,
		},
		{
			FunName: "升级替换介质包",
			Func:    c.Service.ReplaceMedium,
		},
		{
			FunName: "重启proxy",
			Func:    c.Service.RestartProxy,
		},
		{
			FunName: "refresh backend",
			Func:    c.Service.RefreshBackends,
		},
	}
	if err := steps.Run(); err != nil {
		return err
	}
	logger.Info("完成本地升级: 成功")
	return nil
}
